README.md
# color [![Build Status](https://travis-ci.org/harthur/color.svg?branch=master)](https://travis-ci.org/harthur/color)

> JavaScript library for color conversion and manipulation with support for CSS color strings.

```js
var color = Color("#7743CE");

color.alpha(0.5).lighten(0.5);

console.log(color.hslString());  // "hsla(262, 59%, 81%, 0.5)"
```

## Install

### browser
Download the latest [color.js](https://github.com/harthur/color/tree/gh-pages). The `Color` object is exported.

### node
For [node](http://nodejs.org) with [npm](http://npmjs.org):

```console
$ npm install color
```

And use with `var Color = require("color")`

## API

### Setters

```js
var color = Color("rgb(255, 255, 255)")
var color = Color({r: 255, g: 255, b: 255})
var color = Color().rgb(255, 255, 255)
var color = Color().rgb([255, 255, 255])
```
Pass any valid CSS color string into `Color()` or a hash of values. Also load in color values with `rgb()`, `hsl()`, `hsv()`, `hwb()`, and `cmyk()`.

```js
color.red(120)
```
Set the values for individual channels with `alpha`, `red`, `green`, `blue`, `hue`, `saturation` (hsl), `saturationv` (hsv), `lightness`, `whiteness`, `blackness`, `cyan`, `magenta`, `yellow`, `black`

### Getters


```js
color.rgb()       // {r: 255, g: 255, b: 255}
```
Get a hash of the rgb values with `rgb()`, similarly for `hsl()`, `hsv()`, and `cmyk()`

```js
color.rgbArray()  // [255, 255, 255]
```
Get an array of the values with `rgbArray()`, `hslArray()`, `hsvArray()`, and `cmykArray()`.

```js
color.red()       // 255
```
Get the value for an individual channel.

### CSS Strings

```js
color.hslString()  // "hsl(320, 50%, 100%)"
```

Different CSS String formats for the color are on `hexString`, `rgbString`, `percentString`, `hslString`, `hwbString`, and `keyword` (undefined if it's not a keyword color). `"rgba"` and `"hsla"` are used if the current alpha value of the color isn't `1`.

### Luminosity

```js
color.luminosity();  // 0.412
```
The [WCAG luminosity](http://www.w3.org/TR/WCAG20/#relativeluminancedef) of the color. 0 is black, 1 is white.

```js
color.contrast(Color("blue"))  // 12
```
The [WCAG contrast ratio](http://www.w3.org/TR/WCAG20/#contrast-ratiodef) to another color, from 1 (same color) to 21 (contrast b/w white and black).

```js
color.light();  // true
color.dark();   // false
```
Get whether the color is "light" or "dark", useful for deciding text color.

### Manipulation

```js
color.negate()         // rgb(0, 100, 255) -> rgb(255, 155, 0)

color.lighten(0.5)     // hsl(100, 50%, 50%) -> hsl(100, 50%, 75%)
color.darken(0.5)      // hsl(100, 50%, 50%) -> hsl(100, 50%, 25%)

color.saturate(0.5)    // hsl(100, 50%, 50%) -> hsl(100, 75%, 50%)
color.desaturate(0.5)  // hsl(100, 50%, 50%) -> hsl(100, 25%, 50%)
color.greyscale()      // #5CBF54 -> #969696

color.whiten(0.5)      // hwb(100, 50%, 50%) -> hwb(100, 75%, 50%)
color.blacken(0.5)     // hwb(100, 50%, 50%) -> hwb(100, 50%, 75%)

color.clearer(0.5)     // rgba(10, 10, 10, 0.8) -> rgba(10, 10, 10, 0.4)
color.opaquer(0.5)     // rgba(10, 10, 10, 0.8) -> rgba(10, 10, 10, 1.0)

color.rotate(180)      // hsl(60, 20%, 20%) -> hsl(240, 20%, 20%)
color.rotate(-90)      // hsl(60, 20%, 20%) -> hsl(330, 20%, 20%)

color.mix(Color("yellow"))        // cyan -> rgb(128, 255, 128)
color.mix(Color("yellow"), 0.3)   // cyan -> rgb(77, 255, 179)

// chaining
color.green(100).greyscale().lighten(0.6)
```

### Clone

You can can create a copy of an existing color object using `clone()`:

```js
color.clone() // -> New color object
```

And more to come...

## Propers

The API was inspired by [color-js](https://github.com/brehaut/color-js). Manipulation functions by CSS tools like Sass, LESS, and Stylus.


CHANGELOG.md
# 0.11.0 - 2016-01-02

- Fixed: Hue modulo when specifying values < 0 or > 360
([#76](https://github.com/MoOx/color/pull/76))
- Corrected some package.json information

# 0.10.1 - 2015-07-02

- Fixed: handling of 0% mix
([#65](https://github.com/harthur/color/pull/65))

# 0.10.0 - 2015-07-02

- Fixed: `mix()` works with a 0..1 range (instead of 0..100 since 0.9.0)
([#64](https://github.com/harthur/color/pull/64))

# 0.9.0 - 2015-06-21

- Fixed: `mix()` implementation is now the same as in Sass
([#60](https://github.com/harthur/color/pull/60))

# 0.8.0 - 2015-03-03

- Removed: bower support
- Removed: component(1) support
- Changed: Upgrade to color-string 0.3

---

Check out commit logs for older releases


src/color.js
/* MIT license */
(function() {

    var convert = require("color-convert"),
        string = require("color-string");

    var Color = function(obj) {
        if (obj instanceof Color) return obj;
        if (!(this instanceof Color)) return new Color(obj);

        this.values = {
            rgb: [0, 0, 0],
            hsl: [0, 0, 0],
            hsv: [0, 0, 0],
            hwb: [0, 0, 0],
            cmyk: [0, 0, 0, 0],
            alpha: 1
        }

        // parse Color() argument
        if (typeof obj == "string") {
            var vals = string.getRgba(obj);
            if (vals) {
                this.setValues("rgb", vals);
            } else if (vals = string.getHsla(obj)) {
                this.setValues("hsl", vals);
            } else if (vals = string.getHwb(obj)) {
                this.setValues("hwb", vals);
            } else {
                throw new Error("Unable to parse color from string \"" + obj + "\"");
            }
        } else if (typeof obj == "object") {
            var vals = obj;
            if (vals["r"] !== undefined || vals["red"] !== undefined) {
                this.setValues("rgb", vals)
            } else if (vals["l"] !== undefined || vals["lightness"] !== undefined) {
                this.setValues("hsl", vals)
            } else if (vals["v"] !== undefined || vals["value"] !== undefined) {
                this.setValues("hsv", vals)
            } else if (vals["w"] !== undefined || vals["whiteness"] !== undefined) {
                this.setValues("hwb", vals)
            } else if (vals["c"] !== undefined || vals["cyan"] !== undefined) {
                this.setValues("cmyk", vals)
            } else {
                throw new Error("Unable to parse color from object " + JSON.stringify(obj));
            }
        }
    }

    Color.prototype = {
        rgb: function(vals) {
            return this.setSpace("rgb", arguments);
        },
        hsl: function(vals) {
            return this.setSpace("hsl", arguments);
        },
        hsv: function(vals) {
            return this.setSpace("hsv", arguments);
        },
        hwb: function(vals) {
            return this.setSpace("hwb", arguments);
        },
        cmyk: function(vals) {
            return this.setSpace("cmyk", arguments);
        },

        rgbArray: function() {
            return this.values.rgb;
        },
        hslArray: function() {
            return this.values.hsl;
        },
        hsvArray: function() {
            return this.values.hsv;
        },
        hwbArray: function() {
            if (this.values.alpha !== 1) {
                return this.values.hwb.concat([this.values.alpha])
            }
            return this.values.hwb;
        },
        cmykArray: function() {
            return this.values.cmyk;
        },
        rgbaArray: function() {
            var rgb = this.values.rgb;
            return rgb.concat([this.values.alpha]);
        },
        hslaArray: function() {
            var hsl = this.values.hsl;
            return hsl.concat([this.values.alpha]);
        },
        alpha: function(val) {
            if (val === undefined) {
                return this.values.alpha;
            }
            this.setValues("alpha", val);
            return this;
        },

        red: function(val) {
            return this.setChannel("rgb", 0, val);
        },
        green: function(val) {
            return this.setChannel("rgb", 1, val);
        },
        blue: function(val) {
            return this.setChannel("rgb", 2, val);
        },
        hue: function(val) {
            return this.setChannel("hsl", 0, val);
        },
        saturation: function(val) {
            return this.setChannel("hsl", 1, val);
        },
        lightness: function(val) {
            return this.setChannel("hsl", 2, val);
        },
        saturationv: function(val) {
            return this.setChannel("hsv", 1, val);
        },
        whiteness: function(val) {
            return this.setChannel("hwb", 1, val);
        },
        blackness: function(val) {
            return this.setChannel("hwb", 2, val);
        },
        value: function(val) {
            return this.setChannel("hsv", 2, val);
        },
        cyan: function(val) {
            return this.setChannel("cmyk", 0, val);
        },
        magenta: function(val) {
            return this.setChannel("cmyk", 1, val);
        },
        yellow: function(val) {
            return this.setChannel("cmyk", 2, val);
        },
        black: function(val) {
            return this.setChannel("cmyk", 3, val);
        },

        hexString: function() {
            return string.hexString(this.values.rgb);
        },
        rgbString: function() {
            return string.rgbString(this.values.rgb, this.values.alpha);
        },
        rgbaString: function() {
            return string.rgbaString(this.values.rgb, this.values.alpha);
        },
        percentString: function() {
            return string.percentString(this.values.rgb, this.values.alpha);
        },
        hslString: function() {
            return string.hslString(this.values.hsl, this.values.alpha);
        },
        hslaString: function() {
            return string.hslaString(this.values.hsl, this.values.alpha);
        },
        hwbString: function() {
            return string.hwbString(this.values.hwb, this.values.alpha);
        },
        keyword: function() {
            return string.keyword(this.values.rgb, this.values.alpha);
        },

        rgbNumber: function() {
            return (this.values.rgb[0] << 16) | (this.values.rgb[1] << 8) | this.values.rgb[2];
        },

        luminosity: function() {
            // http://www.w3.org/TR/WCAG20/#relativeluminancedef
            var rgb = this.values.rgb;
            var lum = [];
            for (var i = 0; i < rgb.length; i++) {
                var chan = rgb[i] / 255;
                lum[i] = (chan <= 0.03928) ? chan / 12.92 : Math.pow(((chan + 0.055) / 1.055), 2.4)
            }
            return 0.2126 * lum[0] + 0.7152 * lum[1] + 0.0722 * lum[2];
        },

        contrast: function(color2) {
            // http://www.w3.org/TR/WCAG20/#contrast-ratiodef
            var lum1 = this.luminosity();
            var lum2 = color2.luminosity();
            if (lum1 > lum2) {
                return (lum1 + 0.05) / (lum2 + 0.05)
            };
            return (lum2 + 0.05) / (lum1 + 0.05);
        },

        level: function(color2) {
            var contrastRatio = this.contrast(color2);
            return (contrastRatio >= 7.1) ? 'AAA' : (contrastRatio >= 4.5) ? 'AA' : '';
        },

        dark: function() {
            // YIQ equation from http://24ways.org/2010/calculating-color-contrast
            var rgb = this.values.rgb,
                yiq = (rgb[0] * 299 + rgb[1] * 587 + rgb[2] * 114) / 1000;
            return yiq < 128;
        },

        light: function() {
            return !this.dark();
        },

        negate: function() {
            var rgb = []
            for (var i = 0; i < 3; i++) {
                rgb[i] = 255 - this.values.rgb[i];
            }
            this.setValues("rgb", rgb);
            return this;
        },

        lighten: function(ratio) {
            this.values.hsl[2] += this.values.hsl[2] * ratio;
            this.setValues("hsl", this.values.hsl);
            return this;
        },

        darken: function(ratio) {
            this.values.hsl[2] -= this.values.hsl[2] * ratio;
            this.setValues("hsl", this.values.hsl);
            return this;
        },

        saturate: function(ratio) {
            this.values.hsl[1] += this.values.hsl[1] * ratio;
            this.setValues("hsl", this.values.hsl);
            return this;
        },

        desaturate: function(ratio) {
            this.values.hsl[1] -= this.values.hsl[1] * ratio;
            this.setValues("hsl", this.values.hsl);
            return this;
        },

        whiten: function(ratio) {
            this.values.hwb[1] += this.values.hwb[1] * ratio;
            this.setValues("hwb", this.values.hwb);
            return this;
        },

        blacken: function(ratio) {
            this.values.hwb[2] += this.values.hwb[2] * ratio;
            this.setValues("hwb", this.values.hwb);
            return this;
        },

        greyscale: function() {
            var rgb = this.values.rgb;
            // http://en.wikipedia.org/wiki/Grayscale#Converting_color_to_grayscale
            var val = rgb[0] * 0.3 + rgb[1] * 0.59 + rgb[2] * 0.11;
            this.setValues("rgb", [val, val, val]);
            return this;
        },

        clearer: function(ratio) {
            this.setValues("alpha", this.values.alpha - (this.values.alpha * ratio));
            return this;
        },

        opaquer: function(ratio) {
            this.setValues("alpha", this.values.alpha + (this.values.alpha * ratio));
            return this;
        },

        rotate: function(degrees) {
            var hue = this.values.hsl[0];
            hue = (hue + degrees) % 360;
            hue = hue < 0 ? 360 + hue : hue;
            this.values.hsl[0] = hue;
            this.setValues("hsl", this.values.hsl);
            return this;
        },

        mix: function(color2, weight) {
            weight = 1 - (weight == null ? 0.5 : weight);

            // algorithm from Sass's mix(). Ratio of first color in mix is
            // determined by the alphas of both colors and the weight
            var t1 = weight * 2 - 1,
                d = this.alpha() - color2.alpha();

            var weight1 = (((t1 * d == -1) ? t1 : (t1 + d) / (1 + t1 * d)) + 1) / 2;
            var weight2 = 1 - weight1;

            var rgb = this.rgbArray();
            var rgb2 = color2.rgbArray();

            for (var i = 0; i < rgb.length; i++) {
                rgb[i] = rgb[i] * weight1 + rgb2[i] * weight2;
            }
            this.setValues("rgb", rgb);

            var alpha = this.alpha() * weight + color2.alpha() * (1 - weight);
            this.setValues("alpha", alpha);

            return this;
        },

        toJSON: function() {
            return this.rgb();
        },

        clone: function() {
            return new Color(this.rgb());
        }
    }


    Color.prototype.getValues = function(space) {
        var vals = {};
        for (var i = 0; i < space.length; i++) {
            vals[space.charAt(i)] = this.values[space][i];
        }
        if (this.values.alpha != 1) {
            vals["a"] = this.values.alpha;
        }
        // {r: 255, g: 255, b: 255, a: 0.4}
        return vals;
    }

    Color.prototype.setValues = function(space, vals) {
        var spaces = {
            "rgb": ["red", "green", "blue"],
            "hsl": ["hue", "saturation", "lightness"],
            "hsv": ["hue", "saturation", "value"],
            "hwb": ["hue", "whiteness", "blackness"],
            "cmyk": ["cyan", "magenta", "yellow", "black"]
        };

        var maxes = {
            "rgb": [255, 255, 255],
            "hsl": [360, 100, 100],
            "hsv": [360, 100, 100],
            "hwb": [360, 100, 100],
            "cmyk": [100, 100, 100, 100]
        };

        var alpha = 1;
        if (space == "alpha") {
            alpha = vals;
        } else if (vals.length) {
            // [10, 10, 10]
            this.values[space] = vals.slice(0, space.length);
            alpha = vals[space.length];
        } else if (vals[space.charAt(0)] !== undefined) {
            // {r: 10, g: 10, b: 10}
            for (var i = 0; i < space.length; i++) {
                this.values[space][i] = vals[space.charAt(i)];
            }
            alpha = vals.a;
        } else if (vals[spaces[space][0]] !== undefined) {
            // {red: 10, green: 10, blue: 10}
            var chans = spaces[space];
            for (var i = 0; i < space.length; i++) {
                this.values[space][i] = vals[chans[i]];
            }
            alpha = vals.alpha;
        }
        this.values.alpha = Math.max(0, Math.min(1, (alpha !== undefined ? alpha : this.values.alpha)));
        if (space == "alpha") {
            return;
        }

        // cap values of the space prior converting all values
        for (var i = 0; i < space.length; i++) {
            var capped = Math.max(0, Math.min(maxes[space][i], this.values[space][i]));
            this.values[space][i] = Math.round(capped);
        }

        // convert to all the other color spaces
        for (var sname in spaces) {
            if (sname != space) {
                this.values[sname] = convert[space][sname](this.values[space])
            }

            // cap values
            for (var i = 0; i < sname.length; i++) {
                var capped = Math.max(0, Math.min(maxes[sname][i], this.values[sname][i]));
                this.values[sname][i] = Math.round(capped);
            }
        }
        return true;
    }

    Color.prototype.setSpace = function(space, args) {
        var vals = args[0];
        if (vals === undefined) {
            // color.rgb()
            return this.getValues(space);
        }
        // color.rgb(10, 10, 10)
        if (typeof vals == "number") {
            vals = Array.prototype.slice.call(args);
        }
        this.setValues(space, vals);
        return this;
    }

    Color.prototype.setChannel = function(space, index, val) {
        if (val === undefined) {
            // color.red()
            return this.values[space][index];
        }
        // color.red(100)
        this.values[space][index] = val;
        this.setValues(space, this.values[space]);
        return this;
    }

     window.Color = module.exports = Color;

})();
